server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -410,6 +441,204 @@ public void clusterChanged(ClusterChangedEvent event) {
                 }
             }
         }
+        if (currentMaster == null && clusterService.localNode().isMasterNode()) {
+            /*
+             * This begins polling all master-eligible nodes for cluster formation information. However there's a 10-second delay before it
+             * starts, so in the normal situation where during a master transition it flips from master1 -> null -> master2, it the
+             * polling tasks will be canceled before any requests are actually made.
+             */
+            beginPollingClusterFormationInfo();
+        } else {
+            cancelPollingClusterFormationInfo();
+        }
+    }
+
+    private void beginPollingClusterFormationInfo() {
+        cancelPollingClusterFormationInfo();
+        clusterFormationInfoTasks = getMasterEligibleNodes().stream()
+            .map(masterNode -> beginPollingClusterFormationInfo(masterNode, clusterFormationResponses))
+            .collect(Collectors.toList());
+    }
+
+    private void cancelPollingClusterFormationInfo() {
+        clusterFormationInfoTasks.forEach(Scheduler.Cancellable::cancel);
+        /*
+         * Recreates the map so that we don't read old information, or worse get stuck with information about a node that has been
+         * removed from the cluster.
+         */
+        clusterFormationResponses = new ConcurrentHashMap<>();
+    }
+
+    Scheduler.Cancellable beginPollingClusterFormationInfo(
+        // Non-private for testing
+        DiscoveryNode node,
+        final ConcurrentMap<DiscoveryNode, ClusterFormationStateOrException> nodeToClusterFormationStateMap
+    ) {
+        return new PollClusterFormationStateTask(node, nodeToClusterFormationStateMap).pollUntilCancelled();
+    }
+
+    /*
+     * This inner class wraps the logic of polling a master-eligible node for its cluster formation information (which is needed in the
+     * event that the cluster cannot elect a master node).
+     */
+    // Non-private for testing
+    class PollClusterFormationStateTask {
+        /**
+         * The node that is being polled
+         */
+        private final DiscoveryNode node;
+        /**
+         * This is a reference to the global nodeToClusterFormationStateMap that was current at the time this object was constructed. The
+         * global map is recreated whenever the task is cancelled. Having this reference prevents accidental writes to that map after
+         * cancellation.
```

Comment:
This is a bit confusing as it implies we're copying it (ie "that was current at the time this object was constructed"), but also "references" it ... which means it's going to change in line with the global map. 

I'm also not sure about the accidental writes to "that" map (it's "this" map isn't it :) ?) nor what's accidental about the said writes.

Could we maybe talk about the why we have this map here as opposed to the implementation details of how it arrived here? (apologies if I'm misunderstanding its purpose) 

Update: maybe we should drop `PollClusterFormationStateTask` altogether, in which case ignore the above :)

Reply:
Below is a scenario showing why I'm copying the `clusterFormationResponses` reference to and passing it in rather than using the global `clusterFormationResponses`. Maybe I'm being unnecessarily cautious.

We begin polling master eligible nodes A, B, C
  We create a new global clusterFormationResponses map
  We update the global clusterFormationResponses with results from A, B
We cancel polling
We begin polling master eligible nodes  A, B, D (note that it's a different set)
  We create a new global clusterFormationResponses map
  We update the global clusterFormationResponses with results from C (the write was happening before cancel took effect and I don't think that it will interrupt a map insert but maybe I'm wrong)
  We update the global clusterFormationResponses with results from A, B, D

So now we have result for A, B, C, and D, even though C is no longer part of the cluster. So we get weird results when we look to see if we can form a quorum, or if C knows about the existence of D. Or even if it's not as extreme as a node being replaced, we have results that might be out of date, causing confusion. My simple way to avoid worrying about all of this was to just pass the method a reference to the clusterFormationResponses map that it is supposed to be using.